hoomd/hpmc/ClangCompiler.cc

```diff
@@ -91,8 +92,18 @@ std::unique_ptr<llvm::Module> ClangCompiler::compileCode(const std::string& code
                                                 &diagnostic_printer,
                                                 false);
 
-    // determine the clang resource path
-    std::string resource_path(HOOMD_LLVM_INSTALL_PREFIX);
+    // Store the LLVM installation prefix in a volatile char array so that "/bin/clang" can be added
+    // at runtime instead of compile time. When it is added at compile time, conda rewrites an
+    // embedded string "$BUILD_PREFIX/bin/clang" with "$INSTALL_PREFIX", truncating the string.
+    volatile const char* llvm_install_prefix = HOOMD_LLVM_INSTALL_PREFIX;
+
+    // standard string methods do not accept volatile char arrays, extract the string manually
+    std::string clang_exec_with_path;
+    for (volatile const char* c = llvm_install_prefix; *c != 0; ++c)
+        {
+        clang_exec_with_path.push_back(*c);
+        }
```

Comment:
Is the volatile declaration to prevent any compile-time removal of this code? Do we know the volatile declaration is needed here (and in the above variable) to prevent errors?

Reply:
`volatile` is to provide absolute certainty that the compiler can't remove the variable at compile time. It was smart enough to convert `std::string(LLVM_INSTALL_PREFIX) + "/bin/clang"` into its own constant at compile time. It may or may not be smart enough to convert this for loop - but the `volatile` forces it to consider the array as run time variable: https://en.cppreference.com/w/cpp/language/cv

> Every access (read or write operation, member function call, etc.) made through a glvalue expression of volatile-qualified type is treated as a visible side-effect for the [purposes of optimization](https://en.cppreference.com/w/cpp/language/as_if) (that is, within a single thread of execution, **volatile accesses cannot be optimized out** ...